flight/PiOS/Common/pios_mpu.c

```diff
@@ -0,0 +1,1090 @@
+			accel_data.x = -accel_y;
+			accel_data.z = -accel_z;
+			gyro_data.y  = -gyro_x;
+			gyro_data.x  = -gyro_y;
+			gyro_data.z  = -gyro_z;
+#ifdef PIOS_INCLUDE_MPU_MAG
+			mag_data.x   = -mag_x;
+			mag_data.y   = -mag_y;
+			mag_data.z   = mag_z;
+#endif // PIOS_INCLUDE_MPU_MAG
+			break;
+		case PIOS_MPU_TOP_270DEG:
+			accel_data.y = accel_y;
+			accel_data.x = -accel_x;
+			accel_data.z = -accel_z;
+			gyro_data.y  = gyro_y;
+			gyro_data.x  = -gyro_x;
+			gyro_data.z  = -gyro_z;
+#ifdef PIOS_INCLUDE_MPU_MAG
+			mag_data.x   = mag_y;
+			mag_data.y   = -mag_x;
+			mag_data.z   = mag_z;
+#endif // PIOS_INCLUDE_MPU_MAG
+			break;
+		case PIOS_MPU_BOTTOM_0DEG:
+			accel_data.y = -accel_x;
+			accel_data.x = accel_y;
+			accel_data.z = accel_z;
+			gyro_data.y  = -gyro_x;
+			gyro_data.x  = gyro_y;
+			gyro_data.z  = gyro_z;
+#ifdef PIOS_INCLUDE_MPU_MAG
+			mag_data.x   = mag_x;
+			mag_data.y   = -mag_y;
+			mag_data.z   = -mag_z;
+#endif // PIOS_INCLUDE_MPU_MAG
+			break;
+
+		case PIOS_MPU_BOTTOM_90DEG:
+			accel_data.y = -accel_y;
+			accel_data.x = -accel_x;
+			accel_data.z = accel_z;
+			gyro_data.y  = -gyro_y;
+			gyro_data.x  = -gyro_x;
+			gyro_data.z  = gyro_z;
+#ifdef PIOS_INCLUDE_MPU_MAG
+			mag_data.x   = -mag_y;
+			mag_data.y   = -mag_x;
+			mag_data.z   = -mag_z;
+#endif // PIOS_INCLUDE_MPU_MAG
+			break;
+
+		case PIOS_MPU_BOTTOM_180DEG:
+			accel_data.y = accel_x;
+			accel_data.x = -accel_y;
+			accel_data.z = accel_z;
+			gyro_data.y  = gyro_x;
+			gyro_data.x  = -gyro_y;
+			gyro_data.z  = gyro_z;
+#ifdef PIOS_INCLUDE_MPU_MAG
+			mag_data.x   = -mag_x;
+			mag_data.y   = mag_y;
+			mag_data.z   = -mag_z;
+#endif // PIOS_INCLUDE_MPU_MAG
+			break;
+
+		case PIOS_MPU_BOTTOM_270DEG:
+			accel_data.y = accel_y;
+			accel_data.x = accel_x;
+			gyro_data.y  = gyro_y;
+			gyro_data.x  = gyro_x;
+			gyro_data.z  = gyro_z;
+			accel_data.z = accel_z;
+#ifdef PIOS_INCLUDE_MPU_MAG
+			mag_data.x   = mag_y;
+			mag_data.y   = mag_x;
+			mag_data.z   = -mag_z;
+#endif // PIOS_INCLUDE_MPU_MAG
+			break;
+		}
+
+		int16_t raw_temp = (int16_t)(mpu_rec_buf[IDX_TEMP_OUT_H] << 8 | mpu_rec_buf[IDX_TEMP_OUT_L]);
+		float temperature;
+		if (mpu_dev->mpu_type == PIOS_MPU6500 || mpu_dev->mpu_type == PIOS_MPU9250)
+			temperature = 21.0f + ((float)raw_temp) / 333.87f;
+		else
+			temperature = 35.0f + ((float)raw_temp + 512.0f) / 340.0f;
+
+		// Apply sensor scaling
+		float accel_scale = PIOS_MPU_GetAccelScale();
+		accel_data.x *= accel_scale;
+		accel_data.y *= accel_scale;
+		accel_data.z *= accel_scale;
+		accel_data.temperature = temperature;
+
+		float gyro_scale = PIOS_MPU_GetGyroScale();
+		gyro_data.x *= gyro_scale;
+		gyro_data.y *= gyro_scale;
+		gyro_data.z *= gyro_scale;
+		gyro_data.temperature = temperature;
+
+		PIOS_Queue_Send(mpu_dev->accel_queue, &accel_data, 0);
+		PIOS_Queue_Send(mpu_dev->gyro_queue, &gyro_data, 0);
+
+#ifdef PIOS_INCLUDE_MPU_MAG
+		if (mpu_dev->use_mag) {
+			uint8_t st1 = mpu_rec_buf[IDX_MAG_ST1];
```

Comment:
A good idea for a future extension/PR, and probably a bit of a refactor of the HMC driver to avoid duplicating all that code? At this stage we have no hardware utilizing the MPU I2C bus in this way.